### ISSUE TITLE
Clickhouse Support

### ISSUE THREAD

**Issue by bojdell** [kyleconroy/sqlc#1628]
### What do you want to change?

I've enjoyed using sqlc with postgres. However, I'm starting to do more work with [Clickhouse](https://github.com/ClickHouse/ClickHouse) and it would be great to also use sqlc there. I imagine this is a similar ask to https://github.com/kyleconroy/sqlc/issues/161 for sqlite and the core work required is integrating a query parser for ClickHouse. I don't have much experience in this area, but here is some potentially related content that may be helpful:
* https://github.com/ClickHouse/ClickHouse/issues/7595
* https://github.com/ClickHouse/ClickHouse/pull/11298

**Comment by bojdell** [comment]
@kyleconroy was there an outcome of the triaging?

**Comment by bojdell** [comment]
An alternate solution that's likely easier to implement is to add an argument to bypass schema validation - aka allow query generation without statically checking the queries against the schema and let the user rely on test codepaths to ensure correctness of the logic. Obviously using the argument would sacrifice safety but gain flexibility, enabling support for any database/sql driver. Thoughts?

**Comment by cqhung1412** [comment]
Hi, @kyleconroy I'd like to ask if there is any progress on this.

**Comment by nexovec** [comment]
Are you guys considering this?

**Comment by matthewshirley** [comment]
I was also curious, but I don't think it is possible with the provided ANTLR4, as the output for Go is unusable:

* https://github.com/ClickHouse/ClickHouse/issues/31650
  * https://github.com/antlr/antlr4/pull/3299
 
Support was also dropped for the ANTLR4 files:

* https://github.com/ClickHouse/ClickHouse/pull/25942

Seems that a similar pattern to Postgres could be used with this project:

* https://github.com/AfterShip/clickhouse-sql-parser

**Comment by w0rp** [comment]
I'm very much interested in having this as a feature, as I'd like to use this library to build tools that use ClickHouse as the only SQL-like database.

**Comment by 1saifj** [comment]
Sqlc supporting Clickhouse is a big addition!.

**Comment by adambenhassen** [comment]
Greetings, @kyleconroy! was there any progress on this?

**Comment by seungjulee** [comment]
bump for more support! i would love to also contribute

**Comment by 1saifj** [comment]
support clickhouse!

**Comment by daviddavid533** [comment]
Clickhouse support! 🙌

**Comment by blackopsrepl** [comment]
Clickhouse has a mysql compatible interface:

https://clickhouse.com/docs/interfaces/mysql

Maybe it's possible to do something starting from there. I'll do some tests and report back.

**Comment by nexovec** [comment]
@blackopsrepl There is also a postgres interface:

https://clickhouse.com/docs/interfaces/postgresql

I personally use postgres with sqlc and it's also more used in general, and it's also(I haven't actually measured) faster thanks to pgx. Wouldn't that be the logical choice for interop with clickhouse instead?

**Comment by blackopsrepl** [comment]
@nexovec True, I didn't notice, since last quick I was just working on an extremely quick prototype. Since I mostly use postgres myself, I'll definitely go that route. Thanks!

**Comment by kkggg55** [comment]
need that

**Comment by J404J** [comment]
Clickhouse support! 🙌

**Comment by ShubhamRasal** [comment]
need this

**Comment by stsepelin** [comment]
Clickhouse support! 🫰

**Comment by mgilbir** [comment]
Given that the work had stalled on the other PR, I've opened PR #4220 that adds ClickHouse support. 

We've been using it in production for a month and it works well for our queries. If you try it out, let me know if you hit any problems. I'd be happy to get feedback and figure out how to land this.

**Comment by kyleconroy** [comment]
I've made a first pass at a Clickhouse SQL parser here https://github.com/sqlc-dev/doubleclick

Please try it out! If it can't parse something, please open an issue.

**Comment by vito** [comment]
If anyone wants to carry the baton, I started integrating Doubleclick into #4220 but had to put it on ice for now. Pushed my WIP here: https://github.com/vito/sqlc/commit/6cf786672f654bf18b17298029ff1a9e0f4c4bc4 (disclaimer: heavily Claude driven)

**Comment by mnafees** [comment]
Hey @kyleconroy 👋🏽 Eager to see if there is any progress here? PR https://github.com/sqlc-dev/sqlc/pull/4220 seems like it might be waiting for a review. Happy to help get this over the line if you need some bandwidth and help!

**Comment by andyndang** [comment]
Looks like the PR got closed because of merge conflicts? Is there any plan to revisit it?

**Comment by kyleconroy** [comment]
I won't be revisiting that specific pull request, but Clickhouse support is closer than ever. If you'd like to try out a very early version, you can build sqlc from main and use the new `sqlc parse` and `sqlc analyze` commands. I'll be wiring it up to code gen soon.
